### PR TITLE
Chronos: TSPipeline supports input data creator

### PR DESCRIPTION
**Pull request by liangs6212**

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 1]
```diff
@@ -82,13 +82,25 @@ def evaluate(self, data, metrics=['mse'], multioutput="uniform_average", batch_s
...
-        yhat = self._tsdataset_unscale(yhat)
-        # unscale
-        y = self._tsdataset_unscale(y)
+        if isinstance(data, TSDataset):
+            x, y = self._tsdataset_to_numpy(data, is_predict=False)
+            yhat = self._best_model.inference(torch.from_numpy(x),
+                                              batch_size=batch_size,
+                                              backend=None).numpy()
+            yhat = self._tsdataset_unscale(yhat)
+            # unscale
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:91]: move this comment to line 90

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 2]
```diff
@@ -82,13 +82,25 @@ def evaluate(self, data, metrics=['mse'], multioutput="uniform_average", batch_s
...
+            y = self._tsdataset_unscale(y)
+        elif isinstance(data, types.FunctionType):
+            x, y = TSPipeline._transform_dataloader(data(self._best_config))
+            if self._best_config['batch_size']:
+                batch_size = self._best_config['batch_size']
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:96]: this batch_size should be user defined batch_size

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 3]
```diff
@@ -82,13 +82,25 @@ def evaluate(self, data, metrics=['mse'], multioutput="uniform_average", batch_s
...
+        if isinstance(data, TSDataset):
+            x, y = self._tsdataset_to_numpy(data, is_predict=False)
+            yhat = self._best_model.inference(torch.from_numpy(x),
+                                              batch_size=batch_size,
+                                              backend=None).numpy()
+            yhat = self._tsdataset_unscale(yhat)
+            # unscale
+            y = self._tsdataset_unscale(y)
+        elif isinstance(data, types.FunctionType):
+            x, y = TSPipeline._transform_dataloader(data(self._best_config))
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:94]: data(self._best_config.update({"batch_size":batch_size}))

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 4]
```diff
@@ -82,13 +82,25 @@ def evaluate(self, data, metrics=['mse'], multioutput="uniform_average", batch_s
...
+            if self._best_config['batch_size']:
+                batch_size = self._best_config['batch_size']
+            yhat = self._best_model.inference(torch.from_numpy(x),
+                                              batch_size=batch_size,
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:98]: delete this line

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 5]
```diff
@@ -127,36 +150,58 @@ def predict(self, data, batch_size=32):
...
-                                          backend=None)
-        yhat = self._tsdataset_unscale(yhat)
+        if isinstance(data, TSDataset):
+            x, _ = self._tsdataset_to_numpy(data, is_predict=True)
+            yhat = self._best_model.inference(torch.from_numpy(x),
+                                              batch_size=batch_size,
+                                              backend=None)
+            yhat = self._tsdataset_unscale(yhat)
+        elif isinstance(data, types.FunctionType):
+            x, _ = TSPipeline._transform_dataloader(data(self._best_config), is_predict=True)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:167]: concat y

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 6]
```diff
@@ -187,17 +232,27 @@ def fit(self,
...
+        valid_loader = None
+        if isinstance(data, TSDataset):
+            if batch_size is None:
+                batch_size = self._best_config["batch_size"]
+            train_loader = self._tsdataset_to_loader(data, batch_size=batch_size)
+            if validation_data:
+                valid_loader = self._tsdataset_to_loader(validation_data, batch_size=batch_size)
+        elif isinstance(data, types.FunctionType):
+            if batch_size is None:
+                batch_size = self._best_config["batch_size"]
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:244]: delete line 243-244

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 7]
```diff
@@ -82,13 +83,28 @@ def evaluate(self, data, metrics=['mse'], multioutput="uniform_average", batch_s
...
+                                              backend=None).numpy()
+            # unscale
+            yhat = self._tsdataset_unscale(yhat)
+            y = self._tsdataset_unscale(y)
+        elif isinstance(data, types.FunctionType):
+            yhat_list, y_list = [], []
+            self._best_config.update({'batch_size': batch_size})
+            for x, y in data(self._best_config):
+                yhat = self._best_model.inference(x,
+                                                  backend=None).numpy()
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:99]: We don't need a new line here.

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 8]
```diff
@@ -187,17 +241,27 @@ def fit(self,
...
+            if validation_data:
+                valid_loader = self._tsdataset_to_loader(validation_data, batch_size=batch_size)
+        elif isinstance(data, types.FunctionType):
+            if batch_size:
+                self._best_config.update({'batch_size': batch_size})
+            train_loader = data(self._best_config)
+            if validation_data:
+                valid_loader = validation_data(self._best_config)
+        else:
+            raise RuntimeError("We only support input TSDataset or Function, "
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:258]: or data creator

File: python/chronos/test/bigdl/chronos/autots/test_tspipeline.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+import pytest
+import torch
+
+from torch.utils.data import TensorDataset, DataLoader
+from bigdl.chronos.autots import AutoTSEstimator, TSPipeline
+from bigdl.orca.common import init_orca_context, stop_orca_context
+
+
+def train_data_creator(config):
+    return DataLoader(TensorDataset(torch.randn(10000,
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/autots/test_tspipeline.py:28]: 10000 might be a little too much for ut? 1000 should be fine

File: python/chronos/test/bigdl/chronos/autots/test_tspipeline.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+    def setUp(self) -> None:
+        pass
+
+    def tearDown(self) -> None:
+        pass
+
+    def test_seq2seq_tsppl_seq2seq_support_dataloader(self):
+        tmp_seq2seq_dir = tempfile.TemporaryDirectory()
+        init_orca_context(cores=4, memory="4g")
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/autots/test_tspipeline.py:56]: I think we need to set `init_ray_on_spark=True`? Can we pass the ut with this unset?

**Review by TheaperDeng — APPROVED** [review]
LGTM, please run jenkins

**Comment by liangs6212** [comment]
http://10.112.231.51:18888/job/BigDL-Chronos-PR-Validation/212/
http://10.112.231.51:18888/job/BigDL-PRVN-chronos-Python-Spark-3.1-py37-ray-part2/212/

**Pull request merged**